### PR TITLE
Add Faster R-CNN and Mask R-CNN

### PR DESCRIPTION
**Pull request by fmassa**
This PR adds support for Faster R-CNN R-50 FPN and Mask R-CNN R-50 FPN.
It also includes reference implementation for training and evaluating those models on COCO, with a distributed evaluation paradigm.

The trained models match reported accuracies when using the same base network (R-50 from Caffe2). When using resnet50 from torchvision, detection results are 0.5 mAP lower for the 1x schedule, as noted by [mmdetection](https://github.com/open-mmlab/mmdetection/blob/master/MODEL_ZOO.md#performance).

Note that these models are NOT backwards-compatible with Detectron or `maskrcnn_benchmark`, as several simplifications were made, in particular:
- the anchor generation has been greatly simplified
- we use torchvision resnet-style

Furthermore, this implementation is 25%+ faster than `maskrcnn_benchmark` during training (for Mask R-CNN R-50 FPN), when using aspect ratio grouping (not currently present in this PR for simplicity).

Reference implementation models and more documentation will be added in a follow-up PR.

### CODE REVIEW

**Comment by fmassa** [comment]
Merging this, I'll be adding models and doc in a follow-up PR as this PR is already too big.

**Comment by karanchahal** [comment]
Thanks so much for this amazing work @fmassa ! 

Will you be adding support for focal loss also ? 

Sorry for this seemingly foolish question but I've been noticing that mask rcnn is a preferred model for detection. Although going by simplicity, wouldn't the retina net model be easier to understand while giving similar accuracies? Is there any reasoning behind preferring the mask rcnn model over the retina net ?

**Comment by fmassa** [comment]
@karanchahal Yes, we plan to add RetinaNet and focal loss in a follow-up next release.

The reason why we added Faster / Mask R-CNN is because of its importance in the CV community, as many research papers have been built on top of it.
But indeed, having simpler models is definitely something I'll be looking into adding into torchvision.

**Comment by mkielak** [comment]
There is problem with AnchorGenerator when using it parallel. The cause of the problem is in     `def cached_grid_anchors(self, grid_sizes, strides):`
cache key should contain device id. In existing implementation you can get tensor from forward() on completely different gpu than image_list and feature maps.
cached_grid_anchors can be implemented as:
```
def cached_grid_anchors(self, grid_sizes, strides, device):
    key = tuple(grid_sizes) + tuple(strides) + (str(device),)
    if key in self._cache:
        return self._cache[key]
    anchors = self.grid_anchors(grid_sizes, strides)
    self._cache[key] = anchors
     return anchors
```
and call inside forward method:
```anchors_over_all_feature_maps = self.cached_grid_anchors(grid_sizes, strides, feature_maps[0].device)```

**Comment by fmassa** [comment]
@mkielak can you open a new issue about this?

**Comment by mkielak** [comment]
After adding this comment, I realized that issues will be better place for it and I opened new issue: https://github.com/pytorch/vision/issues/1255
